README.md
# IP Blacklist Plugin

The **IP Blacklist** Plugin is an extension for [Grav CMS](http://github.com/getgrav/grav). This plugin detects abusive requests and blocks abusive IPs. Additionally, it interfaces with [AbuseIPDB](https://www.abuseipdb.com/) for blacklisting and reporting.

> This plugin operates only within Grav, and is not a suitable replacement for a server firewall. It _may_ help reduce the impact of malicious bots on your site performance and bandwidth use, but it can only do so much. I'm not a security expert, so please don't take my word for anything.

Already familiar with Grav plugins? [__Skip to Usage__](#usage)!

---

## Installation

Installing the IP Blacklist plugin can be done in one of three ways: The admin method lets you quickly install the plugin via the Admin Plugin, The GPM (Grav Package Manager) installation method lets you do so with a simple terminal command, and the manual method lets you do so via a zip file.

### Admin Plugin (Preferred)

If you use the Admin Plugin, you can install the plugin directly by browsing the `Plugins`-menu and clicking on the `Add` button.

### GPM Installation

To install the plugin via the [GPM](http://learn.getgrav.org/advanced/grav-gpm), through your system's terminal (also called the command line), navigate to the root of your Grav-installation, and enter:

    bin/gpm install ip-blacklist

This will install the IP Blacklist plugin into your `/user/plugins`-directory within Grav. Its files can be found under `/your/site/grav/user/plugins/ip-blacklist`.

### Manual Installation

To install the plugin manually, download the zip-version of this repository and unzip it under `/your/site/grav/user/plugins`. Then rename the folder to `ip-blacklist`. You can find these files on [GitHub](https://github.com/aricooperdavis/grav-plugin-ip-blacklist) or via [GetGrav.org](http://getgrav.org/downloads/plugins#extras).

You should now have all the plugin files under

    /your/site/grav/user/plugins/ip-blacklist
	
> NOTE: This plugin is a modular component for Grav which may require other plugins to operate, please see its [blueprints.yaml-file on GitHub](https://github.com/aricooperdavis/grav-plugin-ip-blacklist/blob/master/blueprints.yaml).

---

## Configuration

Before configuring this plugin, you should copy the `user/plugins/ip-blacklist/ip-blacklist.yaml` to `user/config/plugins/ip-blacklist.yaml` and only edit that copy.

Here is the default configuration and an explanation of available options:

```yaml
enabled: true
# Enable the overall plugin
abuseipdb_key: null
# Your AbuseIPDB API key, required for fetching the AbuseIPDB blacklist or reporting abusive IPs to AbuseIPDB
logging: false
# Whether to log filtering/blacklisting actions to the Grav debug log (useful for troubleshooting custom filters)
enable_blacklisting: false
# Whether to block IPs that appear on the blacklist
sources:
# Which blacklists to search incoming IPs for
  local: false
  # The local blacklist updated by filtering (see below)
  abuseipdb: false
  # The AbuseIPDB blacklist (API Key required, see above)
enable_auto_cache: false
# Whether to use the Grav Scheduler to cache the AbuseIPDB blacklist daily (recommended)
enable_auto_clean: false
# Whether to use the Grav Scheduler to trim the size of the local blacklist daily
auto_clean_len: 10000
# The max length to which the local blacklist is trimmed by the Auto-Clean job
response: '400'
# The HTTP response code to return to blacklisted requests (options: 400, 403, 418, 503)
enable_reporting: false
# Whether to report abusive IPs to AbuseIPDB (API Key required, see above)
enable_filtering: false
# Whether to filter incoming requests in order to detect abuse and add the abusive IPs to the local blacklist. Overrides individual filter `enabled` attribute.
filters: |
# A list of regex patterns against which incoming request URIs are matched to detect abusive behaviour
  # Enter one regex pattern per line. Disable patterns with a comment (#). E.g.:
  # /path/to/exploit\.php/?action=hack
  /_ignition/execute-solution
  [...]
```

Note that if you use the Admin Plugin, a file with your configuration named ip-blacklist.yaml will be saved in the `user/config/plugins/`-folder once the configuration is saved in the Admin.

---

## Usage

This plugin has 2 main features that go hand in hand (but may be used independently):

* __Blacklisting__ (blocking) abusive IPs from your site
* __Filtering__ incoming requests to detect abusive ones

Both of these features integrate well with the Freemium [AbuseIPDB](https://www.abuseipdb.com) service; you can use their list of abusive IPs for blacklisting, or report abusive IPs detected by your filtering to their database.

### Blacklisting

Blacklisting works by checking the IP from which a request to your site originates against a list of known abusive IPs. If the IP is found on a blacklist then a plain HTTP response code is sent in return, and no further Grav processing occurs, saving you valuable server resources.

![A diagram illustrating that visitors to your site from blacklisted IP addresses are sent an error response and your site does not process the request.](./resources/blacklisting.svg "Visitors to your site from blacklisted IP addresses are sent an error response and your site does not process the request.")

You can customise the response code sent to blacklisted IPs in the plugin configuration. The response codes available to you are:

| __Response Code__ | __Response Message__ |
| :- | :- |
| 400 | [Bad Request](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/400) |
| 403 (default) | [Forbidden](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/403) |
| 418 | [I'm a teapot](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/418) |
| 503 | [Service Unavailable](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/503) |

You can choose to use a local blacklist, or the AbuseIPDB blacklist, or both.

> __Note:__ Using the AbuseIPDB blacklist requires a free [AbuseIPDB API key](https://www.abuseipdb.com/account/api).

If you choose to use the AbuseIPDB blacklist then we recommend enabling the `Enable AbuseIPDB Auto-Cache` option in the plugin configuration to cache the AbuseIPDB blacklist in the background using the Grav Scheduler. This will prevent the first site visitor of the day having to wait for Grav to cache the AbuseIPDB blacklist before the page loads.

If you choose to use a local blacklist then you will need to add IPs to it yourself. This is easy if you enable the Filtering feature, which detects abusive requests to your site and adds the responsible IPs to your local blacklist.

You can also add IPs to your local blacklist using the IP Blacklist page in the Admin plugin. This is a nice graphical interface for the `local` table of the `/user/data/ip-blacklist/blacklists.sqlite` database.

![Screenshot of the IP Blacklist Admin Interface, showing how the local blacklist can be searched, and IP addresses added and removed.](./resources/IP%20Blacklist%20Admin%20Interface.png "IP addresses can be added and removed from the local blacklist from the Admin interface.")

If you wish to limit the size of the local blacklist you can enable the Auto-Clean Grav Scheduler job and set a maximum number of IP addresses in the plugin configuration. This job will run daily to trim the local blacklist to size and compress the database.

### Filtering
Filtering works by matching the [URI](https://en.wikipedia.org/wiki/Uniform_Resource_Identifier#Example_URIs) of incoming requests to your site against known abusive requests. This is done using regex pattern matching, and this plugin comes by default with a number of regex patterns to match known exploits.

![A diagram illustrating that requests for paths that are known to be abusive are caught and the IP responsible is added to your blacklist.](./resources/filtering.svg "Requests for paths that are known to be abusive are caught and the IP responsible is added to your blacklist.")

Once an abusive request has been identified, the IP is recorded and added to the local blacklist. You may also choose for these abusive IPs to automatically be reported to AbuseIPDB.

> __Note:__ Reporting abusive IPs to AbuseIPDB requires _both_ a free [AbuseIPDB API key](https://www.abuseipdb.com/account/api) _and_ the elevated [Reporting Privilege](https://www.abuseipdb.com/account/request-reporting-privilege).

You can customise and extend the filters used to identify abusive requests via the plugin configuration. The `filters` config value is just a multiline string where each line is a regex filter. You can add, modify, and delete lines, or comment them out with a `#`. Be aware that all regex characters need escaping.

Be careful doing this - they're regex patterns and matching is performed against the entire URI, so you'll need to escape any regex symbols i.e. `?` should be `\?` and `.co.uk` should be `\.co\.uk`.

![Screenshot of the IP Blacklist Plugin Configuration, showing how filters can be customised.](./resources/IP%20Blacklist%20Plugin%20Configuration.png "Filters can be customised via the plugin configuration.")

As these filters are stored in your plugin configuration they will not be automatically updated when the plugin is updated. As such a notification will be shown in the IP Blacklist admin interface when an update to the filters is available, which will enable you to automatically or manually install the new default filters.

---

## Stats

We have the plugin running on one of our sites, set up to report all abusive IPs to AbuseIPDB:

<a href="https://www.abuseipdb.com/user/102256" title="AbuseIPDB is an IP address blacklist for webmasters and sysadmins to report IP addresses engaging in abusive behavior on their networks">
	<img src="https://www.abuseipdb.com/contributor/102256.svg" alt="AbuseIPDB Contributor Badge" style="width: 391px;background: #35c246 linear-gradient(rgba(255,255,255,0), rgba(255,255,255,.3) 50%, rgba(0,0,0,.2) 51%, rgba(0,0,0,0));padding: 5px;">
</a>

---

## FAQs
__Q: Does IP blacklisting work if my site uses Cloudflare?__

A: Yes, when Cloudflare passes on a request to your server it adds the originating IP to the request headers. If these headers are present we check use these to check against the blacklist rather than the apparent IP of the request (which would be Cloudflare's own servers).

Note that Cloudflare may choose to serve a cached copy of your website to a blacklisted IP, and this cannot be prevented, but it will have no negative impact on your server.

__Q: When I use the AbuseIPDB blacklist my site sometimes takes ages to load!__

A: Enable the Grav Scheduler `AbuseIPDB Cache` task, which will cache the AbuseIPDB blacklist in the background.

Why is this needed? To keep your site loading quickly we cache as much of the AbuseIPDB blacklist as their API allows, and refresh this cache every 24 hours. If your site receives a request and the blacklist hasn't been cached yet, or the cache is out of date, then it may take a few seconds to fetch it from the AbuseIPDB servers.

__Q: Some IPs that are listed on AbuseIPDB are not being blacklisted!__

A: The AbuseIPDB API only allows us to fetch a blacklisting containing the IPs that they are most confident are abusive. As such, if the IP you're testing isn't in that top range then it may not be blocked. The number of IPs that the API allows us to cache depends on your AbuseIPDB account:

| __Standard__ | __Basic Subscription__ | __Premium Subscription__ |
| :- | :- | :- |
| 10,000 | 100,000 | 500,000 |

For the time being you may add abusive IPs to your local database manually, but we're working on a feature that allows IPs that aren't blacklisted to be checked retrospectively against the AbuseIPDB API `check` endpoint and added to our local blacklist if they're known to be abusive by AbuseIPDB - stay tuned!

<!-- You may wish to enable the `Check individual IPs` option in the plugin configuration. This checks non-blacklisted IPs individually against the AbuseIPDB API after the request has happened. If the IP is found to be blacklisted on AbuseIPDB then it is added to your local blacklist so that future requests can be intercepted.

Whilst this API endpoint will return details about any IP on the blacklist, it is limited to a certain number of requests per day. You can increase the number of requests per day that you may make by verifying that you are a webmaster, and even further by displaying a supporter badge on your site:

| __Basic Account__ | __Verified Webmaster__ | __Verified Supporter__ |
| :- | :- | :- |
| 1,000 | 3,000 | 5,000 |
 -->

__Q: Some IPs are getting blacklisted that are not abusive!__

A: We have no control over the AbuseIPDB blacklist, so if you believe that an IP has been added to it by mistake then you must request that it is taken down directly with AbuseIPDB (for which a free account is required).

If you are using the local blacklist then you can remove the IP from your blacklist in the IP Blacklist section of the Admin plugin. This is a nice graphical interface for the `local` table of the `/user/data/ip-blacklist/blacklists.sqlite` database.

__Q: Some of the filters are catching legitimate requests, whilst some abusive requests aren't being filtered at all!__

A: We try to ensure that the default filters shipped with the plugin cover a decent range of abusive requests but don't match common non-abusive URIs. You may need to tweak these filters a little for your particular setup.

You can modify the filters in the `filters` option of the plugin configuration (see above for details).

__Q: Help, I was testing the filters and have blacklisted my own IP!__

A: You silly goose! You'll have to remove your IP from the `local` table in the `/user/data/ip-blacklist/blacklists.sqlite` database file, because you will be locked out of the Admin plugin. Your IP is only blacklisted from your site, not your server, so you'll be able to log-in and make this change using ssh, SFTP, or cPanel.

If you aren't comfortable with manipulating the database then you can just delete the whole file and it'll be re-created (empty!) when needed.

__Q: The plugin doesn't work on my site!__

A: Some shared hosting providers are quite aggressive with their security practices. You can usually get in contact with them and ask them to relax some of their security measures for your user, but knowing which ones are responsible for breaking the plugin can be tricky.

<!--
Here is a table of some symptoms and causes that may be useful for identifying what's breaking the plugin:

| Symptom | Cause |
| :- | :- |
| The admin interface shows loading spinners. | ? |
-->

If you need help getting it fixed please [open an Issue](https://github.com/aricooperdavis/grav-plugin-ip-blacklist/issues) and we'll try to help. This will also allow us to modify the plugin to avoid these security measures, and we can add your fix to the documentation to help others.

---

## To Do

If you would like to help us develop this plugin then please consider starting with some of these projects that we've got on the go:

- [ ] Add lots more default filters!

CHANGELOG.md
# v1.0.7
## 06-02-2023

1. [](#new)
    * Added `logging` config option to log filtering/blacklisting actions to help troubleshoot custom filters.

# v1.0.6
## 31-01-2023

1. [](#improved)
	* Added more filter patterns for common exploits

# v1.0.5
## 25-01-2023

1. [](#new)
    * Added notification in admin interface when plugin has been updated but filters have not.

# v1.0.4
## 23-01-2023

1. [](#improved)
    * Changed format of `filters` config to string to improve admin interface useability.

# v1.0.3
##  20-01-2023

1. [](#bugfix)
    * Fixed bug where URI being passed to filters did not include file extension in some cases.

# v1.0.2
##  20-01-2023

1. [](#new)
    * Added auto-clean job to limit local blacklist size.
2. [](#improved)
    * Added more filter patterns for common exploits.

# v1.0.1
##  19-01-2023

1. [](#new)
    * Added stats panel to admin interface.
2. [](#improved)
    * Use European date format in changelog.

# v1.0.0
##  12-01-2023

1. [](#new)
    * ChangeLog started...


ip-blacklist.php
<?php
namespace Grav\Plugin;

use Composer\Autoload\ClassLoader;
use Grav\Common\Grav;
use Grav\Common\Plugin;
use Grav\Common\Filesystem\Folder;
use Grav\Framework\Psr7\Response;
use RocketTheme\Toolbox\Event\Event;
use Symfony\Component\Yaml\Yaml;

use GuzzleHttp\Client;
use SQLite3;

/**
 * Class IPBlacklistPlugin
 * @package Grav\Plugin
 */
class IPBlacklistPlugin extends Plugin
{

    public static $db;

    /**
     * @return array
     *
     * The getSubscribedEvents() gives the core a list of events
     *     that the plugin wants to listen to. The key of each
     *     array section is the event that the plugin listens to
     *     and the value (in the form of an array) contains the
     *     callable (or function) as well as the priority. The
     *     higher the number the higher the priority.
     */
    public static function getSubscribedEvents(): array
    {
        return [
            'onPluginsInitialized' => [
                // Uncomment following line when plugin requires Grav < 1.7
                // ['autoload', 100000],
                ['onPluginsInitialized', 0]
            ]
        ];
    }

    /**
     * Composer autoload
     *
     * @return ClassLoader
     */
    public function autoload(): ClassLoader
    {
        return require __DIR__ . '/vendor/autoload.php';
    }

    /**
     * Initialize the plugin
     */
    public function onPluginsInitialized(): void
    {
        // Enable the main events we are interested in
        $this->enable([
            'onRequestHandlerInit' => ['onRequestHandlerInit', 0],
            'onSchedulerInitialized' => ['onSchedulerInitialized', 0],
        ]);

        // Only load admin pages if required
        if (!$this->isAdmin() || !$this->grav['user']->authenticated) {
            return;
        }

        $this->enable([
            'onAdminMenu' => ['onAdminMenu', 0],
            'onAdminTwigTemplatePaths' => ['onAdminTwigTemplatePaths', 0],
            'onAdminAfterSave' => ['onAdminAfterSave',0],
        ]);

        return;
    }

    /**
     * Handle admin page requests, and filter requests if desired
     */
    public function onRequestHandlerInit($request, $handler)
    {
        // Handle admin page requests
        $path = $request->getRoute()->getRoute();

        if (
            $this->isAdmin()
            && $this->grav['user']->authenticated
            && $path == '/admin/ip-blacklist/data'
        ) {
            $body = (array) $request->getRequest()->getParsedBody();
            if (!isset($body['action'])) {
                $response = new Response(400, [], json_encode(['Error' => 'No action']));
                $request->setResponse($response);
                return;
            }

            // Make appropriate DB query
            $db = $this->getDatabase();
            switch ($body['action']) {
                case 'update':
                    $this->updateFilterList();
                    $data = 1;
                    break;
    
                case 'hide':
                    $this->logFilterListUpdate($this->getReleaseVersion());
                    $data = 1;
                    break;
                
                case 'filter-list':
                    // Check if filter list is up-to-date
                    $release = $this->getReleaseVersion();
                    $stmt = $db->prepare('SELECT ("updated") FROM "updated" WHERE "table" = "filters" LIMIT 1');
                    $result = $stmt->execute();
                    $data = [
                        "installed" => $result->fetchArray(SQLITE3_NUM)[0],
                        "release" => $release,
                    ];
                    break;

                case 'last-25':
                    // Get last 25 IPs added to local blacklist
                    $stmt = $db->prepare('SELECT ("ip") FROM "local" ORDER BY "rowid" DESC LIMIT 25');
                    $result = $stmt->execute();
                    $data = [];
                    while ($row = $result->fetchArray(SQLITE3_NUM)) {
                        array_push($data, $row[0]);
                    }
                    break;

                case 'stats':
                    // Get stats from local blacklist
                    $stmt = $db->prepare('SELECT COUNT(1) FROM "local"');
                    $result = $stmt->execute();
                    $data = [
                        "Number of IPs" => $result->fetchArray(SQLITE3_NUM)[0],
                        "Size on Disk" => $this->human_filesize(filesize(
                                $this->grav['locator']->findResource('user://data/ip-blacklist/blacklists.sqlite', true)
                            )),
                    ];
                    break;

                case 'search':
                    // Search local blacklist for IP
                    $stmt = $db->prepare('SELECT ("ip") FROM "local" WHERE "ip"=:ip LIMIT 1');
                    $stmt->bindValue(':ip', $body['ip'], SQLITE3_TEXT);
                    $result = $stmt->execute();
                    $data = (int)$result->fetchArray(SQLITE3_NUM);
                    break;

                case 'add':
                    // Add IP to local blacklist
                    $stmt = $db->prepare('INSERT INTO "local" ("ip") VALUES (:ip)');
                    $stmt->bindValue(':ip', $body['ip'], SQLITE3_TEXT);
                    $data = (int)(bool)$stmt->execute();
                    break;

                case 'remove':
                    // Remove IP from local blacklist
                    $stmt = $db->prepare('DELETE FROM "local" WHERE ("ip" = :ip)');
                    $stmt->bindValue(':ip', $body['ip'], SQLITE3_TEXT);
                    $data = (int)(bool)$stmt->execute();
                    break;

                default:
                    // Report improper query
                    $response = new Response(400, [], json_encode(['Error' => 'Invalid action']));
                    $request->setResponse($response);
                    return;
            }

            // Parse results and send response
            $response = new Response(200, [], json_encode($data));
            $request->setResponse($response);
            return;
        }

        // Filter incoming requests
        $config = $this->config();
        if (!$config['enable_filtering'] && !$config['enable_blacklisting']) {
            return;
        }
        $ip = $this->getRequestIp();

        // Blacklisting
        if ($config['enable_blacklisting']) {
            if ($this->queryBlacklists($ip)) {
                if ($config['logging']) {
                    $this->grav['log']->debug("[IP Blacklist] Blocked request from blacklisted IP: {$ip}");
                }
                $this->rejectRequest($request);
                return;
            }
        }

        // Filtering
        if ($config['enable_filtering']) {
            $uri = $this->grav['uri']->getCurrentUri();
            $filter = strtok($config['filters'], "\n");
            while ($filter !== false) {
                // Filter for abusive requests
                $filter = trim($filter);
                if (substr($filter, 0, 0) != '#' && preg_match('~'.$filter.'~', $uri)) {
                    if ($config['logging']) {
                        $this->grav['log']->debug("[IP Blacklist] Request matched filter '{$filter}': '{$uri}'");
                    }
                    $this->addIpToBlacklist($ip);
                    if ($config['logging']) {
                        $this->grav['log']->debug("[IP Blacklist] IP blacklisted: '{$ip}'");
                    }
                    if ($config['enable_reporting']) {
                        $this->reportIp($ip, $path);
                        if ($config['logging']) {
                            $this->grav['log']->debug("[IP Blacklist] IP reported: '{$ip}'");
                        }
                    }
                    if ($config['enable_blacklisting']) {
                        $this->rejectRequest($request);
                        if ($config['logging']) {
                            $this->grav['log']->debug("[IP Blacklist] Request rejected: '{$uri}'");
                        }
                        return;
                    }
                    break;
                }
                $filter = strtok("\n");
            }
            if ($config['logging']) {
                $this->grav['log']->debug("[IP Blacklist] Request passed filters: '{$uri}'");
            }
        }
    }

    /**
     * Add auto_cache event to scheduler
     */
    public function onSchedulerInitialized(Event $e)
    {
        $config = $this->config();
        $scheduler = $e['scheduler'];
        if ($config['enable_auto_clean']) {
            $job = $scheduler->addFunction('Grav\Plugin\IPBlacklistPlugin::cleanBlacklists', [], 'ip-blacklist-auto-clean');
            $job->backlink('plugins/ip-blacklist');
            $job->at('35 2 * * *');
        }
        if ($config['enable_auto_cache'] && $config['enable_blacklisting'] && $config['sources']['abuseipdb']) {
            $job = $scheduler->addFunction('Grav\Plugin\IPBlacklistPlugin::updateAbuseipdbBlacklist', [], 'ip-blacklist-auto-cache');
            $job->backlink('plugins/ip-blacklist');
            $job->at('45 2 * * *');
        }
    }

    /**
     * Add page to admin plugin menu
     */
    public function onAdminMenu()
    {
        $options = [
            'hint' => 'IP Blacklist',
            'route' => 'ip-blacklist',
            'icon' => 'fa-ban',
            'authorize' => 'admin.users',
        ];
        $this->grav['twig']->plugins_hooked_nav['IP Blacklist'] = $options;
    }

    /**
     * Add admin page resources to admin twig templates
     */
    public function onAdminTwigTemplatePaths($event)
    {
        if (in_array($this->grav['uri']->route(), ['/admin/ip-blacklist'])) {
            $event['paths'] = array_merge(
                $event['paths'],
                [__DIR__.'/admin/templates']
            );
        }
        return $event;
    }

    /**
     * Detect when plugin config is updated and log plugin version
     */
    public function onAdminAfterSave($event)
    {
        $bp = $event['object']->blueprints();
        if ($bp['slug'] === 'ip-blacklist') {
            $this->logFilterListUpdate($bp['version']);
        }
    }

    /**
     * Add an IP to the local blacklist
     */
    public function addIpToBlacklist(string $ip)
    {
        $db = $this->getDatabase();
        $stmt = $db->prepare('INSERT OR IGNORE INTO "local" ("ip") VALUES (:ip)');
        $stmt->bindValue(':ip', $ip, SQLITE3_TEXT);
        $stmt->execute();
    }

    /**
     * Query blacklists for IP
     */
    public function queryBlacklists(string $ip): bool
    {
        $config = $this->config();
        $db = $this->getDatabase();

        // Ensure AbuseIPDB blacklist is in-date
        if ($config['sources']['abuseipdb']) {
            $result = $db->query('SELECT ("updated") FROM "updated" WHERE "table" = "abuseipdb" LIMIT 1');
            $updated = $result->fetchArray()['updated'];
            if ($updated < time()-86400) {
                $this->updateAbuseipdbBlacklist();
            }
        }

        foreach ($config['sources'] as $source => $enabled) {
            if ($enabled) {
                $stmt = $db->prepare("SELECT 1 FROM {$source} WHERE ip=:ip");
                $stmt->bindValue(':ip', $ip, SQLITE3_TEXT);
                $result = $stmt->execute();
                if (!empty($result->fetchArray())) {
                    return true;
                }
            }
        }

        return false;
    }

    /**
     * Updates the AbuseIPDB blacklist using the AbuseIPDB API
     */
    public static function updateAbuseipdbBlacklist()
    {
        $db = self::getDatabase();
        $config = Grav::instance()['config']->get('plugins.ip-blacklist');
        if (!isset($config['abuseipdb_key'])) {
            Grav::instance()['log']->debug('Cannot update AbuseIPDB blacklist without API key.');
            return;
        }

        // Fetch plaintext blacklist
        $client = new Client([
            'base_uri' => 'https://api.abuseipdb.com/api/v2/'
        ]);
        $response = $client->request('GET', 'blacklist', [
            // TODO: implement customisable blacklist params e.g. exceptCountries, limit, confidenceMinimum
            'query' => [
                'limit' => '500000',
            ],
            'headers' => [
                'Accept' => 'text/plain',
                'Key' => $config['abuseipdb_key'],
            ],
        ]);
        if ($response->getStatusCode() != 200) {
            // TODO: update plugin config to disable AbuseIPDB if api unauthorised
            Grav::instance()['log']->debug('Could not update AbuseIPDB blacklist: '.$response->getBody());
            return;
        }
        $result = $response->getBody();

        // Delete existing rows and add new ones iteratively
        $db->exec('DELETE FROM "abuseipdb"');
        $stmt = $db->prepare('INSERT OR IGNORE INTO "abuseipdb" ("ip") VALUES (:ip)');
        $stmt->bindParam(':ip', $line, SQLITE3_TEXT);
        $line = strtok($result, "\r\n");
        while ($line !== false) {
            $stmt->execute();
            $line = strtok("\r\n");
        }
        // Update updated table entry
        $updated = time();
        $stmt = $db->prepare('INSERT OR REPLACE INTO "updated" ("table","updated") VALUES ("abuseipdb",:updated)');
        $stmt->bindValue(':updated', $updated, SQLITE3_TEXT);
        $stmt->execute();
    }

    /**
     * Report IP to AbuseIPDB
     */
    public function reportIp(string $ip, string $path)
    {
        $config = $this->config();
        $client = new Client([
            'base_uri' => 'https://api.abuseipdb.com/api/v2/'
        ]);
        $response = $client->request('POST', 'report', [
            'form_params' => [
                'ip' => $ip,
                'categories' => '21',
                'comment' => 'Probe for vulnerabilities. Path attempted: '.$path,
            ],
            'headers' => [
                'Accept' => 'application/json',
                'Key' => $config['abuseipdb_key'],
            ],
        ]);
        if ($response->getStatusCode() == 200) {
            return;
        } else {
            // TODO: change plugin config to disable reporting if unauthorised api response
            $this->grav['log']->debug('AbuseIPDB report failed ('.$ip.'): '.$response->getBody());
        }
    }

    /**
     * Returns a reference to the database, creating and opening it if required
     */
    public static function getDatabase(): SQLite3
    {
        // Return reference to database if already initialised
        if (isset(self::$db)) {
            return self::$db;
        }
        // Get/create plugin data folder
        $data_dir = Grav::instance()['locator']->findResource('user://data', true).'/ip-blacklist';
        if (!file_exists($data_dir)) {
            Folder::create($data_dir);
        }
        // Get/create db and timestamps
        self::$db = new SQLite3($data_dir.'/blacklists.sqlite');
        self::$db->exec('CREATE TABLE IF NOT EXISTS "local" ("ip" TEXT NOT NULL UNIQUE, PRIMARY KEY("ip"))');
        self::$db->exec('CREATE TABLE IF NOT EXISTS "abuseipdb" ("ip" TEXT NOT NULL UNIQUE, PRIMARY KEY("ip"))');
        self::$db->exec('CREATE TABLE IF NOT EXISTS "updated" ("table" TEXT NOT NULL UNIQUE, "updated" TEXT NOT NULL DEFAULT 0, PRIMARY KEY("table"))');
        self::$db->exec('INSERT OR IGNORE INTO "updated" ("table") VALUES ("filters"),("abuseipdb")');

        return self::$db;
    }

    /**
     * Trim local blacklist to size and vacuum database
     */
    public static function cleanBlacklists()
    {
        // Get config details
        $config = Grav::instance()['config']->get('plugins.ip-blacklist');
        $limit = $config['auto_clean_len'];

        // Trim local blacklist
        $db = self::getDatabase();
        $stmt = $db->prepare('DELETE FROM "local" WHERE "rowid" NOT IN (SELECT "rowid" FROM "local" ORDER BY "rowid" DESC LIMIT :limit)');
        $stmt->bindValue(':limit', $limit, SQLITE3_INTEGER);
        $stmt->execute();

        // Vacuum blacklist db
        $db->exec('VACUUM');
    }

    /**
     * Update the filter list with the most recent on GitHub
     */
    public function updateFilterList()
    {
        $client = new Client([
            'base_uri' => 'https://cdn.jsdelivr.net/gh/aricooperdavis/grav-plugin-ip-blacklist/ip-blacklist.yaml'
        ]);
        $response = $client->request('GET');
        if ($response->getStatusCode() !== 200) {
            $this->grav['log']->debug('Could not fetch ip-blacklist.yaml from jsDelivr: '.$response->getBody());
        }

        $yaml = Yaml::parse($response->getBody());
        $this->grav['config']->set('plugins.ip-blacklist.filters', $yaml['filters']);
        $this->saveConfig('ip-blacklist');

        $this->logFilterListUpdate($this->getReleaseVersion());
    }

    /**
     * Log plugin version at which filter list is updated
     */
    public function logFilterListUpdate($version=null)
    {
        $version = $version ?? $this->getBlueprint()['version'] ;

        $db = $this->getDatabase();
        $stmt = $db->prepare('INSERT OR REPLACE INTO "updated" ("table", "updated") VALUES ("filters", :updated)');
        $stmt->bindValue(':updated', $version, SQLITE3_TEXT);
        $stmt->execute();
    }

    /**
     * Get the release version code
     */
    public function getReleaseVersion(): string
    {
        $client = new Client([
            'base_uri' => 'https://data.jsdelivr.com/v1/package/gh/aricooperdavis/grav-plugin-ip-blacklist'
        ]);
        $response = $client->request('GET');
        if ($response->getStatusCode() != 200) {
            $this->grav['log']->debug('Could not get version numbers from jsDelivr: '.$response->getBody());
            return 0;
        }
        return json_decode($response->getBody(), true)['versions'][0];
    }

    /**
     * Reject request with appropriate HTTP response
     */
    public function rejectRequest($request)
    {
        $text = [
            400 => 'Bad Request',
            403 => 'Forbidden',
            418 => 'I\'m a teapot',
            503 => 'Service Unavailable',
        ];
        $config = $this->config();
        $response = new Response($config['response'], [], $text[$config['response']]);
        $request->setResponse($response);
    }

    /**
     * Get the IP of a user, even if they are behind cloudflare
     * https://github.com/francodacosta/grav-plugin-page-stats/blame/47ff58a7de94860244ffe24c2cb82bc83841ce85/page-stats.php#L99
     */
    public function getRequestIp(): string
    {
        if (isset($_SERVER["HTTP_CF_CONNECTING_IP"])) {
            $_SERVER['REMOTE_ADDR'] = $_SERVER["HTTP_CF_CONNECTING_IP"];
            $_SERVER['HTTP_CLIENT_IP'] = $_SERVER["HTTP_CF_CONNECTING_IP"];
        }
        $client  = @$_SERVER['HTTP_CLIENT_IP'];
        $forward = @$_SERVER['HTTP_X_FORWARDED_FOR'];
        $remote  = $_SERVER['REMOTE_ADDR'];

        if (filter_var($client, FILTER_VALIDATE_IP)) {
            $ip = $client;
        } elseif (filter_var($forward, FILTER_VALIDATE_IP)) {
            $ip = $forward;
        } else {
            $ip = $remote;
        }
        return $ip;
    }

    public function human_filesize($bytes, $decimals = 0) {
        $sz = 'BKMGTP';
        $factor = floor((strlen($bytes) - 1) / 3);
        return sprintf("%.{$decimals}f", $bytes / pow(1024, $factor)) . @$sz[$factor];
    }

}